conformance/utils/kubernetes/apply.go

```diff
@@ -41,45 +42,95 @@ import (
 // Applier prepares manifests depending on the available options and applies
 // them to the Kubernetes cluster.
 type Applier struct {
-	NamespaceLabels map[string]string
-	// ValidUniqueListenerPorts maps each listener port of each Gateway in the
-	// manifests to a valid, unique port. There must be as many
-	// ValidUniqueListenerPorts as there are listeners in the set of manifests.
-	// For example, given two Gateways, each with 2 listeners, there should be
-	// four ValidUniqueListenerPorts.
-	// If empty or nil, ports are not modified.
-	ValidUniqueListenerPorts []v1alpha2.PortNumber
+	namespaceLabels map[string]string
+	// availablePorts holds ports that can still be assigned
+	availablePorts PortSet
+	// assignedPorts tracks ports used for Gateway objects
+	assignedPorts map[types.NamespacedName]PortSet
+	portsLock     sync.Mutex
+}
+
+// NewApplier creates a new Applier object.
+// validPorts should contain a list of valid ports for Gateway listeners.
+// There must be as many validPorts as the maximum number of ports
+// used simultaneously.
+// For example, given one Gateway with 2 listeners on the same port and one
+// Gateway with 2 listeners on different ports, there should be at least three
+// validPorts.
+// If empty or nil, Gateway listener ports are not modified.
+func NewApplier(namespaceLabels map[string]string, validPorts []v1alpha2.PortNumber) *Applier {
+	return &Applier{
+		namespaceLabels: namespaceLabels,
+		availablePorts:  validPorts,
+		assignedPorts:   map[types.NamespacedName]PortSet{},
+	}
+}
+
+// freePortsFor is used to mark ports used by a Gateway as available. It does
```

Comment:
Is this intended to be used when a Gateway is being deleted? The name made me think this was just returning a list of available ports for the given input.

Reply:
Yeah that's free as a verb not an adjective. I'll rename